rfcs/20180817-variables-20.md

```diff
@@ -0,0 +1,140 @@
+
+
+## Main changes
+
+The API for Variables will then change in the following ways for TF 2.0:
+
+
+
+*   tf.Variable will become an abstract base class with a well-defined interface and a scoped factory to construct instances
+    *   users will be able to implement their own variable-like objects by subclassing tf.Variable and adding a scoped factory function to use those variables
+*   variable_scope and get_variable will be removed
+    *   the tf 1.0 version of variable_scope and get_variable will be left in tf.compat.v1
+    *   to control variable naming users can use tf.name_scope + tf.Variable
+    *   whether a variable is shared across sessions / processes will be controlled by a constructor argument to tf.Variable; no other type of scope reuse will be done in the framework
+    *   scoped partitioning will be implemented as a factory function at first
+    *   libraries and users are encouraged to reuse variables by reusing their objects, like Keras layers do
+    *   custom_getters will have the following API: [variable_creator_scope](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/variable_scope.py#L2395)
+*   the default implementation of the tf.Variable interface will be ResourceVariable
+    *   RefVariable will be kept in tf.compat.v1 and will be the default implementation for tf.compat.v1.Variable
+    *   tf.compat.v1.Variable will have a use_resource argument to control whether a resource variable or a ref variable will be created
+*   symbols like tf.assign* will be removed in favor of methods in tf.Variable
+    *   in tf.compat.v1 these symbols will be marked as deprecated and will call the corresponding methods in the Variable object instead
+
+
+## Detailed changes
+
+
+### tf.Variable class
+
+The tf.Variable class will be an abstract base class which defines a tf.Variable interface. Initially this interface will have enough abstract methods such that the user-visible API of tf.Variable does not change.
+
+There will be two main implementations of this interface: RefVariable, with the legacy ref edges, available only in tf.compat.v1, and ResourceVariable, which is the default for the v2 API. PartitionedVariable, MirroredVariable, _UnreadVariable, CastVariable, etc, are other implementations which are part of the core library. None of these implementations will be publicly visible, only tf.Variable will be.
+
+Constructing variables is done by calling tf.Variable(*args, **kwargs). Under the hood this will call a hierarchy of scoped constructor functions, similar to what is now done in variable_scope.variable. Each such constructor function can do some combination of:
```

Comment:
1. Could you explain why we choose `tf.Variable(*args, **kwargs)`, rather than `tf.get_variable`, to construct variables? 
    > The tf.Variable class will be an abstract base class which defines a tf.Variable interface.
   
     If `tf.Variable` will be an abstract base class, how to call `tf.Variable(*args, **kwargs)`?

2. Could you explain what is scoped constructor functions?

Reply:
1. tf.get_variable was created to handle silent sharing of variables in the graph. This behavior is being removed.

2. See the link I updated about variable_creator_scope

Reply:
Hi @alextp.
Could you please show an example about how to create a `PartitionedVariable` via API `tf.Variable(*args, **kwargs) ` ? My question is whether user should pass an indicator to show what kinds of concrete `Variable` to create ?  Does it mean the parameters `*args` and `**kwargs` are exposed to users without any limit?

Reply:
> 1. Will it be possible to recover tf.Variable objects only from a graph or graph_def, just like it's now possible to do with tf.Variable.from_proto? We work a lot with managing models restored purely from graph def files, without necessarily having all the code that produced the original graph. The ability to restore basic TF objects such as tf.Variables directly from graph def data only is a must for us.
> 2. How is the above affected by tf.Variable types written by users?
> 3. Will it be possible to explicitly recreate or recover tf.Variable objects from other non-python-object pieces of data like in some way?

+1 on this. It is crucial for us to restore them from serialized graphdefs. Currently we use a `RestoredVariable` class inheriting from `tf.Variable`, but Refvariable changes in TF1.11 are breaking this inheritance. See issues [#23591](https://github.com/tensorflow/tensorflow/issues/23591), [#22648](https://github.com/tensorflow/tensorflow/issues/22648).